### PR TITLE
Robot interchange during game play (Both Div)

### PR DESCRIPTION
**Pull request by leonardoSCosta**
From the Rule proposal:

> We aim for less stoppages and especially with walls, there will probably be much less chances to interchange robots. We already encourage teams to automatically interchange robots by replacing them at the midline, but the game still had to be stopped and even halted in 2019. This wasted overall game time.
> 
> We propose that robots can always be taken in and out from the midline. No need to stop the game. There should not be any action at the midline during the replacement. That could be enforced by requiring that the ball is not near the interchange position.

> In 2019, a lot of time was lost in HALT, while teams interchanged their robots.
> We want to encourage teams to reduce the number of manual interchanges to a minimum and make use of the automatic interchange to save overall game time.
> 
> We propose that manual robot interchange (if the robot is not taken in/out from the midline) always requires a timeout to be taken.
> If a team is out of timeouts and still needs to take one robot out, the team is allowed to take the robot out without a timeout, but can not use any remaining timeout time. It can only remove the robot from the field as fast as possible.
>

### CODE REVIEW

**Comment by leonardoSCosta** [comment]
To keep the consistency with the do not stop for yellow cards rule #10, we've decided that the automatic interchange will also apply to Div B. Note that this isn't something that is mandatory to be used, but, it is something that can help those who want to use it

File: chapters/substitution.adoc [hunk 1]
```diff
@@ -2,13 +2,22 @@
...
-.. Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
-.. Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
+. Robot substitution during game play:
+.. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+.. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs at least one of the following operations in the given order.
+... Put robots in. If the ball is at least 2 meters away from the halfway line, the robot can be placed on the intersection of the halfway line with any touch line without notifying the <<Referee, referee>>.
+... Take robots out. If a robot is at the intersection of the halfway line with any touch line and the ball is at least 2 meters away from the halfway line, the robot can be removed without notifying the <<Referee, referee>>.
```
Comment (MathewMacDougall) [inline chapters/substitution.adoc]: Is there a threshold for how close the robot must me to the intersection of the halfway line and touch line like before? Or in the case of "live" substitutions are robots now expected to be in the nearly exact position?

File: chapters/substitution.adoc [hunk 2]
```diff
@@ -2,13 +2,22 @@
...
+
+. Robot substitution on next stoppage:
+.. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs at least one of the following operations in the given order.
+... Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
```
Comment (MathewMacDougall) [inline chapters/substitution.adoc]: This is a bit of a nitpick, but if the robot handler follows these steps in order they may temporarily exceed the maximum number of robots of the team on the field because one must be added before another can be removed. Perhaps we should clarify the maximum allowed number of robots of the team on the field must not be exceeded _when the substitutions are finished_? Unless this is intended.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: chapters/substitution.adoc (hunk 2 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: I don't think this is intended, I believe that what it means is that the robot handler can perform any of those actions, not necessarily in order. Doesn't make much sense only being able to take a robot out if you put one in. It is the same idea as previous years, there can never be more robots than the allowed for each team, and, there's nothing that prohibits having less robots. I'll fix this.

File: chapters/substitution.adoc (hunk 1 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: I'll specify it better. I think that the robot doesn't need to be exactly on the intersection, what is most important is that the robot is at least on top of the touch line, to avoid any interference with the vision system, and it could be at 30cm ish from the halfway line.

**Review by MathewMacDougall — APPROVED** [review]
:+1:

File: chapters/gamestructure.adoc [hunk 3]
```diff
@@ -72,6 +72,7 @@ Before the start of the match, every team has to designate one robot handler. Th
 .Duties
 * The robot handler helps <<Match Preparation, preparing the match>>.
 * The robot handler asks the referee for <<Timeouts, timeouts>> if necessary.
+* The robot handler can substitute a robot during game play, <<Robot Substitution, substitutes the robot>>.
```
Comment (g3force) [inline chapters/gamestructure.adoc]: that's not a valid sentence, I guess? :D

File: chapters/substitution.adoc [hunk 4]
```diff
@@ -18,6 +30,7 @@ A robot substitution intent can be made by:
 . A <<Robot Handler, robot handler>> by informing the <<Game Controller Operator, game controller operator>> who in turn enters the intent into the <<Game Controller, game controller>>.
 . A team software by sending a request to the <<Game Controller, game controller>>.
 . The <<Game Controller, game controller>> itself if a team exceeds the maximum number of robots (for example after a team receives a <<Yellow Card, yellow>> or <<Red Card, red card>>).
+. *(Division A only)* If a robot needs to be taken out and it is not in the mid line, a timeout needs to be taken in order for the team to remove the robot from the field. If a team is out of timeouts, it will still be able to take the robot out, however, it will not be able to use any remaining timeout time, the robot must be removed from the field as fast as possible.
```
Comment (g3force) [inline chapters/substitution.adoc]: Can we put this into a separate PR? This has a lot more consequences for teams than the rest and would be a candidate for postponement to 2021.

File: chapters/substitution.adoc [hunk 5]
```diff
@@ -2,13 +2,25 @@
...
-.. Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
-.. Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
+. Robot substitution during game play:
+.. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+.. A robot can only be substituted while obeying the following specifications:
+... Is at least on top of any touch line, if possible the robot can also be placed beyond the field boundaries
+... At a distance from the halfway line that must not exceed 30 centimeters
+.. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs the following operations.
+... Put robots in. If the ball is at least 2 meters away from the halfway line, the robot can be substituted without notifying the <<Referee, referee>>.
```
Comment (g3force) [inline chapters/substitution.adoc]: can you try to remove the 2 meters away definition duplication?

File: chapters/substitution.adoc [hunk 6]
```diff
@@ -2,13 +2,25 @@
...
+... Take robots out. If the robot is placed at the previous specified position and the ball is at least 2 meters away from the halfway line, the robot can be removed without notifying the <<Referee, referee>>.
+
+. Robot substitution on next stoppage:
+.. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs the following operations.
+... Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
+... Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
```
Comment (g3force) [inline chapters/substitution.adoc]: With the new rules, "the next opportunity" does not make sense anymore, because you can now add robots whenever you want.

The rule was added to motivate the automatic interchange. With new changes, there is an even higher motivation for this, because you can substitute robots immediately, so I think it is fine to drop this line.

File: chapters/substitution.adoc [hunk 7]
```diff
@@ -2,13 +2,25 @@
 .Definition
 Robots are substituted by the <<Robot Handler, robot handler>> of the respective team. No other team member is allowed to take robots out or put robots in.
 
-. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs at least one of the following operations in the given order.
-.. Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
-.. Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
+. Robot substitution during game play:
+.. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+.. A robot can only be substituted while obeying the following specifications:
+... Is at least on top of any touch line, if possible the robot can also be placed beyond the field boundaries
```
Comment (g3force) [inline chapters/substitution.adoc]: Suggestion: Is at least partially inside the field margin

but double check if the field margin is defined as such. Maybe add it into the field drawings?

**Review by g3force — CHANGES_REQUESTED** [review]
The definition of the interchange has some duplications, both in the new part and with the existing rules.

I find it confusing that for the interchange during stop, there is a 1m radius, while during play there is another rule.

I would remove the old definition with the 1m radius and replace it by the new one, except that during stop, robots can be removed from everywhere on the field.

We can further optimize the procedure:
During stop, robots can be removed.
Robots can always be added (given the conditions for substitution during game).
I think, that way the whole section can hugely be simplified.

File: chapters/substitution.adoc (hunk 6 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: Makes sense, but what if a team still wants to manually replace a robot? For example, a robot that for some reason got a low battery and cannot move, it won't cause any damage to itself or other robots, thus, a emergency stop is not required, but it will still need to be manually removed from the field, this kind of situation can wait until a game stoppage

Comment (leonardoSCosta) [reply chapters/substitution.adoc]: The 1 meter rule doesn't make much sense, since now it is possible to put robots in at any time, so I'll just remove it

File: chapters/substitution.adoc (hunk 4 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: Done. PR #16

File: chapters/substitution.adoc (hunk 6 above)
Comment (g3force) [reply chapters/substitution.adoc]: Without #16, teams can still substitute robots during stop as usual. No need for an emergency stop.

File: chapters/substitution.adoc [hunk 8]
```diff
@@ -2,21 +2,36 @@
...
 Robots are substituted by the <<Robot Handler, robot handler>> of the respective team. No other team member is allowed to take robots out or put robots in.
 
-. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs at least one of the following operations in the given order.
-.. Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
-.. Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
+. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+
+. Robots can always be substituted during game play if the following specifications are respected:
+
+.. Is at least partially inside the <<Field Surface, field margin>>
```
Comment (g3force) [inline chapters/substitution.adoc]: > The robot is as least...

File: chapters/substitution.adoc [hunk 9]
```diff
@@ -2,21 +2,36 @@
...
+.. At a distance from the halfway line that must not exceed 30 centimeters
```
Comment (g3force) [inline chapters/substitution.adoc]: Prefer SI units here, so 0.3m.

What was the rational again for 0.3m? We should consider that there might be both robot handlers on the same side, as we gave no limitations as to where they should be. Last year, we asked them to be both on the opposite side from the GCO.

To me, 1m would sound just as fine.

File: chapters/substitution.adoc [hunk 10]
```diff
@@ -2,21 +2,36 @@
...
+.. The ball must be at least 2 meters away from the halfway line
+
+. If the previous specifications are respected, the <<Robot Handler, robot handler>> is allowed to perform the following operations.
+... Put robots in, without notifying the <<Referee, referee>>.
+... Take robots out, without notifying the <<Referee, referee>>.
+
+. Given that robots can be put in at any time, while respecting the previous specifications, on a game stoppage robots can only be removed.
```
Comment (g3force) [inline chapters/substitution.adoc]: That sounds a bit complicated. What about:

> Additionally, robots can be taken out from any position during game stoppage

File: chapters/substitution.adoc [hunk 11]
```diff
@@ -2,21 +2,36 @@
...
+. The maximum allowed number of robots of the team on the field must not be exceeded.
+
 
 .Usage
 Robots can be substituted for any reason. There is no limit on the number of substitutions.
 
 A robot substitution intent can be made by:
 
 . A <<Robot Handler, robot handler>> by informing the <<Game Controller Operator, game controller operator>> who in turn enters the intent into the <<Game Controller, game controller>>.
-. A team software by sending a request to the <<Game Controller, game controller>>.
```
Comment (g3force) [inline chapters/substitution.adoc]: Why is this removed?

File: chapters/substitution.adoc [hunk 12]
```diff
@@ -2,21 +2,36 @@
...
-.. Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
+. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+
+. Robots can always be substituted during game play if the following specifications are respected:
+
+.. Is at least partially inside the <<Field Surface, field margin>>
+.. At a distance from the halfway line that must not exceed 30 centimeters
+.. The ball must be at least 2 meters away from the halfway line
+
+. If the previous specifications are respected, the <<Robot Handler, robot handler>> is allowed to perform the following operations.
```
Comment (g3force) [inline chapters/substitution.adoc]: Originally, the rules were defined with the order and the maximum number of robots in mind, that's why there were those two options (put in, take out). The idea was: The robot handler can put robots in until the maximum number is reached. Then the robot handler can take robots out again. That way, a team had to play with one robot less until the next stoppage.

Now in the new context, this formulation does not make much sense to me anymore. I would just simplify it to one sentence.

File: chapters/substitution.adoc [hunk 13]
```diff
@@ -2,21 +2,36 @@
...
+... Put robots in, without notifying the <<Referee, referee>>.
+... Take robots out, without notifying the <<Referee, referee>>.
+
+. Given that robots can be put in at any time, while respecting the previous specifications, on a game stoppage robots can only be removed.
+.. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs the following operation in the given order.
```
Comment (g3force) [inline chapters/substitution.adoc]: Merge this with the point before?

File: chapters/substitution.adoc [hunk 14]
```diff
@@ -2,21 +2,36 @@
...
+... The <<Robot Handler, robot handler>> informs the <<Referee, referee>> when done.
+... When both teams finished the robot substitution, the <<Referee, referee>> informs the <<Game Controller Operator, game controller operator>>.
+... The <<Game Controller Operator, game controller operator>> performs a <<Stop, stop>> followed by continuing the game.
+
 . If one of the robots taken out is the keeper, the <<Robot Handler, robot handler>> tells the <<Game Controller Operator, game controller operator>> the id of the robot that takes over the keeper role.
-. The <<Robot Handler, robot handler>> informs the <<Referee, referee>> when done.
-. When both teams finished the robot substitution, the <<Referee, referee>> informs the <<Game Controller Operator, game controller operator>>.
-. The <<Game Controller Operator, game controller operator>> performs a <<Stop, stop>> followed by continuing the game.
+
+. The maximum allowed number of robots of the team on the field must not be exceeded.
```
Comment (g3force) [inline chapters/substitution.adoc]: > ... at any time

**Review by g3force — CHANGES_REQUESTED** [review]

File: chapters/substitution.adoc (hunk 9 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: With 0.3 m there would be a total of 0.6 m (both sides of the halfway line), I believe that's enough for more than one robot being substituted at the same time, but I'm okay with 1 m too

File: chapters/substitution.adoc (hunk 10 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: Yeah, I agree, thanks :sweat_smile:

File: chapters/substitution.adoc (hunk 11 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: With the new automatic procedure I didn't see why this would be necessary. Should it be kept?

Comment (g3force) [reply chapters/substitution.adoc]: if a robot is broken and must be taken off the field, you still want to request substitution.

File: chapters/substitution.adoc [hunk 15]
```diff
@@ -2,13 +2,25 @@
...
+. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+
+. Robots can always be substituted during game play if the following specifications are respected:
+
+.. The robot is at least partially inside the <<Field Surface, field margin>>
+.. The robot is at a distance from the halfway line that must not exceed 1 meter
+.. The ball must be at least 2 meters away from the halfway line
+
+. If the previous specifications are respected, the <<Robot Handler, robot handler>> is allowed to put robots in and take robots out, without notifying the <<Referee, referee>>.
+.. Additionally, robots can be taken out from any position during game stoppage by the <<Robot Handler, robot handler>> using the procedure below:
```
Comment (g3force) [inline chapters/substitution.adoc]: I think there is no need for the extra indents here, as there is only one point on the second level (this line).

Could you btw. add a reference for game stoppage here? :)

File: chapters/substitution.adoc [hunk 16]
```diff
@@ -2,13 +2,25 @@
...
 Robots are substituted by the <<Robot Handler, robot handler>> of the respective team. No other team member is allowed to take robots out or put robots in.
 
-. A <<Robot Handler, robot handler>> who's team is marked for robot substitution performs at least one of the following operations in the given order.
-.. Put robots in. The maximum allowed number of robots of the team on the field must not be exceeded.
-.. Take robots out. If a robot that is taken out is closer than 1 meter to the intersection of the halfway line with one of the touch lines, another robot can be put in right away. Otherwise the robot handler has to wait until the next substitution opportunity.
+. The <<Robot Handler, robot handler>> should prefer to use long sleeves and colors that won't interfere with the vision system.
+
+. Robots can always be substituted during game play if the following specifications are respected:
+
+.. The robot is at least partially inside the <<Field Surface, field margin>>
```
Comment (g3force) [inline chapters/substitution.adoc]: General note: It seems that most of the bullet point lists end with a dot. It would be nice to keep consistency.

File: chapters/substitution.adoc [hunk 17]
```diff
@@ -2,13 +2,25 @@
...
+.. The robot is at a distance from the halfway line that must not exceed 1 meter
+.. The ball must be at least 2 meters away from the halfway line
+
+. If the previous specifications are respected, the <<Robot Handler, robot handler>> is allowed to put robots in and take robots out, without notifying the <<Referee, referee>>.
```
Comment (g3force) [inline chapters/substitution.adoc]: Sry for nitpicking, but this is sentence seems partially redundant. You already said before, that the specifications have to be respected and now you repeat it. I would just add the "without notifying to the point above:

> Robots can always be substituted during game play without notifying the <<Referee, referee>> if the following specifications are respected:

**Review by g3force — CHANGES_REQUESTED** [review]

File: chapters/substitution.adoc (hunk 17 above)
Comment (leonardoSCosta) [reply chapters/substitution.adoc]: No problem, if you find anything that could be improved just tell me

**Review by g3force — APPROVED** [review]
